### DISCUSSION TITLE
Cross-toolchain library linking

### DISCUSSION THREAD

**Discussion by Kaltxi** [microsoft/vcpkg#38466] [General]
Hello!

I've been building some of my projects on windows-x64 with clang, while installing vcpkg packages using default toolchain (which is msvc). But now I've encountered an issue with one of the dependencies (OpenImageIO, and in particular its dependency OpenEXR), which manually adds `\EHsc` flag to cmake and makes my own targets also include it resulting in build failure (clang doesn't understand msvc-styled flags). An example of this can be seen in this issue: https://github.com/microsoft/vcpkg/issues/34926.

The proposed solution (https://github.com/microsoft/vcpkg/discussions/38030) is to build the dependencies with clang also. My question: was I doing it wrong all this time? Is cross-toolchain library linking a bad practice with possibly other problems I'm not aware of, or is it just on per-library basis hurdles like this one with the flag (i.e. if we it compiles, we're good)? Also, if it is the latter - is there a better way for library providers to add compilation flags without breaking builds on foreign toolchain consumers?

**Comment by Neumann-A** [comment]
Use clang-cl? Otherwise you might always encounter some cmake dependency adding build flags the wrong way. OpenEXR should probably do a better compiler check when adding those flags and use a generator expression to disable it.

**Reply by Kaltxi** [reply]
Thanks for your time answering my question! Could you please elaborate on  better compiler check? As far as I understand the problem lies in OpenEXR doing this in their lib setup:

```cmake
if (MSVC)
  set(_openexr_extra_flags "/EHsc")
endif()

if (_openexr_extra_flags)
  target_compile_options(${objlib} PUBLIC ${_openexr_extra_flags})
endif()
```

Since they use public compile option - they require that users to have async exceptions enabled, but that happens as vcpkg builds the package, so we have no info on package user yet. Is there a way to make it more versatile? Maybe so it can work with `-fasync-exceptions`?

**Reply by Neumann-A** [reply]
If those are public flags it would be better to use:
`set(_openexr_extra_flags "$<$<CXX_COMPILER_ID:MSVC>:/EHsc>")`
but there is probably an equivalent for clang since clang-cl also has to support that flag.

**Reply by Kaltxi** [reply]
Thanks for info! I created overlay ports for OpenEXR and Imath (which also has the exact same problem) and it worked! Had no idea generator expressions are passed as is and not evaluated at configure time of the library.
